pkg/trustedresources/verifier/verifier_test.go

```diff
@@ -0,0 +1,419 @@
+							HashAlgorithm: "sha256",
+						},
+					},
+				},
+			},
+		},
+		expectedError: ErrorDecodeKey,
+	}, {
+		name: "from secret error",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Key: &v1alpha1.KeyRef{
+							SecretRef: &v1.SecretReference{
+								Name:      "wrongSecret",
+								Namespace: "wrongNamespace",
+							},
+						},
+					},
+				},
+			},
+		},
+		expectedError: errorSecretNotFound,
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := FromPolicy(context.Background(), fakek8s.NewSimpleClientset(), tc.policy)
+			if !errors.Is(err, tc.expectedError) {
+				t.Errorf("FromPolicy got: %v, want: %v", err, tc.expectedError)
+			}
+		})
+	}
+}
+
+func TestFromKeyRef_Success(t *testing.T) {
+	ctx := context.Background()
+	_, keypath := test.GetSignerFromFile(ctx, t)
+
+	_, pub, err := test.GenerateKeys(pass("1234"))
+	if err != nil {
+		t.Errorf("failed to generate keys: %v", err)
+	}
+	secretData := &v1.Secret{
+		Data: map[string][]byte{
+			"data": pub,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: namespace}}
+
+	k8sclient := fakek8s.NewSimpleClientset(secretData)
+
+	tcs := []struct {
+		name   string
+		keyref string
+	}{{
+		name:   "key in file",
+		keyref: keypath,
+	}, {
+		name:   "key in secret",
+		keyref: fmt.Sprintf("k8s://%s/secret", namespace),
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			verifier, err := fromKeyRef(ctx, tc.keyref, crypto.SHA256, k8sclient)
+			if verifier == nil {
+				t.Errorf("fromKeyRef got empty verifiers")
+			}
+			if err != nil {
+				t.Errorf("fromKeyRef got: %v, want: nil", err)
+			}
+		})
+	}
+}
+
+func TestFromKeyRef_Error(t *testing.T) {
+	ctx := context.Background()
+	_, keypath := test.GetSignerFromFile(ctx, t)
+	tcs := []struct {
+		name          string
+		keyref        string
+		algorithm     crypto.Hash
+		expectedError error
+	}{{
+		name:          "failed to read file",
+		keyref:        "wrongPath",
+		algorithm:     crypto.SHA256,
+		expectedError: ErrorFailedLoadKeyFile,
+	}, {
+		name:          "failed to read from secret",
+		keyref:        fmt.Sprintf("k8s://%s/not-exist-secret", namespace),
+		algorithm:     crypto.SHA256,
+		expectedError: errorSecretNotFound,
+	}, {
+		name:          "failed to read from data",
+		keyref:        keypath,
+		algorithm:     crypto.BLAKE2b_256,
+		expectedError: errorLoadVerifier,
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := fromKeyRef(ctx, tc.keyref, tc.algorithm, fakek8s.NewSimpleClientset())
+			if !errors.Is(err, tc.expectedError) {
+				t.Errorf("fromKeyRef got: %v, want: %v", err, tc.expectedError)
+			}
+		})
+	}
+}
+
+func TestFromSecret_Success(t *testing.T) {
+	_, pub, err := test.GenerateKeys(pass("1234"))
+	if err != nil {
+		t.Errorf("failed to generate keys: %v", err)
```

Comment:
anytime you encounter an error that would make it so the test shouldn't even continue (e.g. in this case the test couldn't generate the keys it needed) you probably want to use `Fatalf` to stop the rest of the test from running (vs Errorf which allows you to accumlate errors as you go and report them all)